balances-tx-payment/rpc/src/lib.rs

```diff
@@ -0,0 +1,194 @@
+};
+//use encointer_rpc::Error;
+use jsonrpsee::{
+	core::{Error as JsonRpseeError, RpcResult},
+	proc_macros::rpc,
+	types::error::{CallError, ErrorCode, ErrorObject},
+};
+pub use pallet_transaction_payment::RuntimeDispatchInfo;
+use pallet_transaction_payment::{FeeDetails, InclusionFee};
+use pallet_transaction_payment_rpc::TransactionPaymentApiServer;
+use sp_api::{Decode, Encode, ProvideRuntimeApi};
+use sp_blockchain::HeaderBackend;
+use sp_core::Bytes;
+use sp_rpc::number::NumberOrHex;
+use sp_runtime::{
+	generic::BlockId,
+	traits::{Block as BlockT, MaybeDisplay},
+};
+use std::sync::Arc;
+
+#[rpc(client, server)]
+pub trait BalancesTxPaymentApi<BlockHash, AssetId>
+where
+	AssetId: 'static + Encode + Decode + Send + Sync,
+{
+	#[method(name = "encointer_queryAssetFeeDetails", blocking)]
+	fn query_asset_fee_details(
+		&self,
+		asset_id: AssetId,
+		encoded_xt: Bytes,
+		at: Option<BlockHash>,
+	) -> RpcResult<FeeDetails<NumberOrHex>>;
+}
+
+pub struct BalancesTxPaymentRpc<C, P, Q, R, S> {
+	client: Arc<C>,
+	transaction_payment: P,
+	_marker: std::marker::PhantomData<Q>,
+	_marker2: std::marker::PhantomData<R>,
+	_marker3: std::marker::PhantomData<S>,
+}
+
+impl<C, P, Q, R, S> BalancesTxPaymentRpc<C, P, Q, R, S> {
+	pub fn new(client: Arc<C>, transaction_payment: P) -> Self {
+		BalancesTxPaymentRpc {
+			client,
+			transaction_payment,
+			_marker: Default::default(),
+			_marker2: Default::default(),
+			_marker3: Default::default(),
+		}
+	}
+}
+
+impl<C, P, Block, AssetId, Balance, AssetBalance>
+	BalancesTxPaymentApiServer<<Block as BlockT>::Hash, AssetId>
+	for BalancesTxPaymentRpc<C, P, Block, Balance, AssetBalance>
+where
+	AssetId: 'static + Clone + Copy + Encode + Decode + Send + Sync + PartialEq,
+	AssetBalance: 'static
+		+ Clone
+		+ Encode
+		+ Decode
+		+ Send
+		+ Sync
+		+ PartialEq
+		+ Into<NumberOrHex>
+		+ MaybeDisplay
+		+ Copy,
+	Block: BlockT,
+	P: TransactionPaymentApiServer<<Block as BlockT>::Hash, RuntimeDispatchInfo<Balance>>,
+	C: ProvideRuntimeApi<Block> + HeaderBackend<Block> + Send + Sync + 'static,
+	C::Api: BalancesTxPaymentApiRuntimeApi<Block, Balance, AssetId, AssetBalance>,
+	Balance:
+		Codec + MaybeDisplay + Copy + TryInto<NumberOrHex> + Send + Sync + 'static + From<u128>,
+{
+	fn query_asset_fee_details(
+		&self,
+		asset_id: AssetId,
+		encoded_xt: Bytes,
+		at: Option<Block::Hash>,
+	) -> RpcResult<FeeDetails<NumberOrHex>> {
+		let api = self.client.runtime_api();
+
+		let balance_fee_details = self.transaction_payment.query_fee_details(encoded_xt, at)?;
+
+		let at = BlockId::hash(at.unwrap_or_else(|| self.client.info().best_hash));
+
+		let try_into_rpc_balance = |value: AssetBalance| {
+			value.try_into().map_err(|_| {
+				JsonRpseeError::Call(CallError::Custom(ErrorObject::owned(
+					ErrorCode::InvalidParams.code(),
+					format!("{} doesn't fit in NumberOrHex representation", value),
+					None::<()>,
+				)))
+			})
+		};
```

Comment:
Closures are nice, but as this closure does not need external context, it would be more readable to extract this into a function.

Reply:
i copied it from here :)
https://github.com/paritytech/substrate/blob/938c0d9363131d862bcc04d2cf913ce9cb9dd5fe/frame/transaction-payment/rpc/src/lib.rs#L146

Reply:
haha, ok.